tests/jdom/tester.ts

```diff
@@ -0,0 +1,233 @@
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+interface ServerDevice {
+    server: JDServiceServer
+    roleName?: string
+}
+
+// Creates a bus with the specified servers (optionally bound to the specified roles),
+// then runs the test function.
+export async function withBus(
+    devices: ServerDevice[],
+    test: (
+        bus: JDBus,
+        serviceMap: Map<JDServiceServer, JDService>
+    ) => Promise<void>
+) {
+    const bus = mkBus()
+
+    // For server devices: add the service provider on the bus and return the device
+    const serverDevices = devices.map(device => {
+        const busDevice = bus.addServiceProvider(
+            new JDServiceProvider([device.server])
+        )
+        return {
+            server: device.server,
+            busDevice: busDevice,
+            roleName: device.roleName,
+        }
+    })
+
+    bus.start()
+
+    // Wait for created devices to be announced, so services become available
+    const serverDeviceIds = serverDevices.map(elt => elt.busDevice.deviceId)
+    await new Promise(resolve => {
+        const devicesIdSet = new Set(serverDeviceIds)
+        const announcedIdSet = new Set()
+        const onHandler = (device: JDDevice) => {
+            if (devicesIdSet.has(device.deviceId)) {
+                announcedIdSet.add(device.deviceId)
+            }
+            if (setEquals(devicesIdSet, announcedIdSet)) {
+                bus.off(DEVICE_ANNOUNCE, onHandler)
+                resolve(undefined)
+            }
+        }
+        bus.on(DEVICE_ANNOUNCE, onHandler)
+    })
+
+    // Bind services to roles
+    // Start by geting server service -> roleName mappings
+    const serverServiceRoles = serverDevices.map(elt => {
+        const services = elt.busDevice.services({
+            serviceClass: elt.server.serviceClass,
+        })
+        assert(
+            services.length > 0,
+            `created device ${elt.busDevice.friendlyName} has no service of ${elt.server.specification.name}`
+        )
+        assert(
+            services.length == 1,
+            `created device ${elt.busDevice.friendlyName} has multiple service of ${elt.server.specification.name}`
+        )
+        return {
+            service: services[0],
+            roleName: elt.roleName,
+        }
+    })
+
+    const roleBindings = serverServiceRoles
+        .filter(
+            elt => !!elt.roleName // filter for where role name is available
+        )
+        .map(elt => {
+            return {
+                role: elt.roleName,
+                serviceClass: elt.service.specification.classIdentifier,
+                service: elt.service,
+            }
+        })
+
+    const roleManager = new RoleManager(bus)
+    roleManager.setRoles(roleBindings)
```

Comment:
do we use the role manager?

Reply:
The basic button test here doesn't, but the adapters code (not part of this PR) does. If you don't think the role manager is generally useful I'll remove it from here?

Reply:
It's fine for this round. We can circle back later on and see if you can componentize this better.